### PR TITLE
vault share pricing and unmatured position valuation

### PR DESCRIPTION
**Pull request by mcclurejt**
fixes https://github.com/delvtech/hyperdrive/issues/1089

This is rfr....

NOTE: The refactoring made some of the existing tests redundant/obsolete. Probably safe to ignore all unit/integration except for `HyperdriveExecution.t.sol`. Since this PR is already huge I'll be redoing those in another pr, and can combine the two before merging.

### CODE REVIEW

File: contracts/interfaces/IEverlong.sol [hunk 1]
```diff
@@ -1,31 +1,46 @@
...
+
+    /// @notice Tracks the total amount of bonds managed by Everlong
+    ///         with the same maturity.
+    struct Position {
+        /// @notice Time when the position matures.
+        uint128 maturityTime;
+        /// @notice Quantity of bonds in the position.
+        uint128 bondAmount;
+        /// @notice Vault share price at time of purchase.
+        uint128 vaultSharePrice;
```
Comment (jalextowle) [inline contracts/interfaces/IEverlong.sol:27]: If this is used for negative interest, we should just use the vault share price from Hyperdrive’s checkpoint. Otherwise, this may not be needed at all. Either way, we don’t need to store it here.

Comment (mcclurejt) [reply contracts/interfaces/IEverlong.sol:27]: Yeah so with negative interest and using the checkpoint vault share price I see a pretty massive difference in estimated vs actual output...

```
 Want: 10239305534593637769923
 Have: 9986496180175740679947
```

However when I use the actual purchase vault share price, I still see a difference, but it's massively less:

```
Want: 9987830847237632187943
Have: 9987830847237632186046
```

Unsure of why this is, maybe I'm pulling the price from the wrong checkpoint?

Comment (mcclurejt) [reply contracts/interfaces/IEverlong.sol]: Now have down to the wei accurate estimates

File: contracts/Everlong.sol [hunk 2]
```diff
@@ -1,22 +1,23 @@
 // SPDX-License-Identifier: Apache-2.0
-pragma solidity 0.8.22;
+pragma solidity ^0.8.20;
```
Comment (jalextowle) [inline contracts/Everlong.sol:2]: What's the rationale for this change? IMO it's good to lock the version for the core contracts and keep the version unlocked for libraries and interfaces that will be used elsewhere.

Comment (mcclurejt) [reply contracts/Everlong.sol:2]: Unintentional. Leftover from debugging a compiler version issue (ended up being from an oz contract upstream import). Will update with pinned version

File: contracts/Everlong.sol [hunk 3]
```diff
@@ -66,64 +67,313 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Storage                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    // ───────────────────────── Immutables ──────────────────────
+
+    // NOTE: Immutables accessed during transactions are left as internal
+    //       to avoid the gas overhead of auto-generated getter functions.
+    //       https://zokyo-auditing-tutorials.gitbook.io/zokyo-gas-savings/tutorials/gas-saving-technique-23-public-to-private-constants
```
Comment (jalextowle) [inline contracts/Everlong.sol:84]: This is interesting. I read through this link, and I actually wrote a test locally to see if there was overhead to accessing `internal` vs `public` immutables. The idea that the cost would be different was surprising to me since we shouldn't be using the public getter when accessing them internally. I didn't find the gas costs to be any different, but I could be missing something.

Here is the `contracts/src/Example.sol`:

```
// SPDX-License-Identifier: Apache-2.0
pragma solidity 0.8.22;

contract ExampleA {
    uint256 public immutable im;

    constructor(uint256 _im) {
        im = _im;
    }

    function add(uint256 _value) external view returns (uint256) {
        return im + _value;
    }
}

contract ExampleB {
    uint256 internal immutable im;

    constructor(uint256 _im) {
        im = _im;
    }

    function add(uint256 _value) external view returns (uint256) {
        return im + _value;
    }
}
```

Here's `test/Example.t.sol`:

```
// SPDX-License-Identifier: Apache-2.0
pragma solidity 0.8.22;

import { console2 as console } from "forge-std/console2.sol";
import { Test } from "forge-std/Test.sol";
import { ExampleA, ExampleB } from "contracts/src/Example.sol";

contract ExampleTest is Test {
    ExampleA internal exampleA;
    ExampleB internal exampleB;

    function setUp() external {
        exampleA = new ExampleA(1);
        exampleB = new ExampleB(1);
    }

    function test_example_a() external {
        uint256 gasBefore = gasleft();
        exampleA.add(10);
        console.log("gas used = %s", gasBefore - gasleft());
    }

    function test_example_b() external {
        uint256 gasBefore = gasleft();
        exampleB.add(10);
        console.log("gas used = %s", gasBefore - gasleft());
    }
}
```

The output when running both tests was:

```
Ran 2 tests for test/Example.t.sol:ExampleTest
[PASS] test_example_a() (gas: 8851)
Logs:
  gas used = 5386

[PASS] test_example_b() (gas: 8829)
Logs:
  gas used = 5386

Suite result: ok. 2 passed; 0 failed; 0 skipped; finished in 5.88ms (1.44ms CPU time)
```

Comment (jalextowle) [reply contracts/Everlong.sol:84]: To be clear, I would definitely expect deployment costs to be higher with the public getters, but assuming that we're going to expose these getters elsewhere, there shouldn't be a difference. The question I was trying to answer was whether or not internal accesses would be more expensive.

File: contracts/Everlong.sol [hunk 4]
```diff
@@ -66,64 +67,313 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
+
+    /// @notice Rebalance the everlong portfolio by closing mature positions
+    ///         and using the proceeds to open new positions.
+    function rebalance() public override {
+        // Close matured positions.
+        _closeMaturedPositions();
+
+        // Spend idle on opening a new position.
+        uint256 toSpend = _asset.balanceOf(address(this));
+        IERC20(_asset).approve(address(_hyperdrive), toSpend);
```
Comment (jalextowle) [inline contracts/Everlong.sol:239]: It would be good to use `forceApprove` here since it's possible for the asset to be a non-standard ERC20 token. It would also be good to add an additional 1 wei of approval so that the storage slot remains warm after the funds are pulled.

File: contracts/Everlong.sol [hunk 5]
```diff
@@ -66,64 +67,313 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
+    function asset() public view override returns (address) {
+        return address(_asset);
+    }
+
+    /// @notice Gets the admin address for Everlong.
+    function admin() external view returns (address) {
+        return _admin;
+    }
+
+    // FIXME: Consider idle liquidity + maybe maxLong?
```
Comment (jalextowle) [inline contracts/Everlong.sol:332]: Is the plan to fix this in this PR or a follow-up?

File: contracts/Everlong.sol [hunk 6]
```diff
@@ -66,64 +67,313 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
 
-    /// @notice Gets whether Everlong uses Hyperdrive's base token to
-    ///         transact.
+    /// @notice Whether Everlong uses Hyperdrive's base token to transact.
+    /// @return Whether Everlong uses Hyperdrive's base token to transact.
     function asBase() external view returns (bool) {
         return _asBase;
     }
+
+    /// @notice Address of the token used to interact with the Hyperdrive instance.
```
Comment (jalextowle) [inline contracts/Everlong.sol:321]: Nit: 

```suggestion
    /// @notice Returns the address of the token used to interact with the Hyperdrive 
    ///                instance.
```

File: contracts/interfaces/IEverlong.sol [hunk 7]
```diff
@@ -1,31 +1,45 @@
...
     IEverlongEvents,
-    IEverlongPositions
+    IEverlongPortfolio
 {
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Structs                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /**
+     * @notice Tracks the total amount of things with stuff.
```
Comment (jalextowle) [inline contracts/interfaces/IEverlong.sol:20]:
```suggestion
     * @notice Contains the information needed to identify an open Hyperdrive position.
```

Comment (jalextowle) [reply contracts/interfaces/IEverlong.sol:20]: The line length may be too long, but I think that's more descriptive.

File: contracts/libraries/HyperdriveExecution.sol [hunk 8]
```diff
@@ -0,0 +1,578 @@
...
+import { IHyperdrive } from "hyperdrive/contracts/src/interfaces/IHyperdrive.sol";
+import { FixedPointMath } from "hyperdrive/contracts/src/libraries/FixedPointMath.sol";
+import { HyperdriveMath } from "hyperdrive/contracts/src/libraries/HyperdriveMath.sol";
+import { SafeCast } from "hyperdrive/contracts/src/libraries/SafeCast.sol";
+import { YieldSpaceMath } from "hyperdrive/contracts/src/libraries/YieldSpaceMath.sol";
+import { Packing } from "openzeppelin/utils/Packing.sol";
+import { IEverlong } from "../interfaces/IEverlong.sol";
+import { IEverlongEvents } from "../interfaces/IEverlongEvents.sol";
+
+// TODO: Extract into its own library.
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:13]: Good call

File: contracts/libraries/HyperdriveExecution.sol [hunk 9]
```diff
@@ -0,0 +1,578 @@
...
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Open Long                                               │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @notice Opens a long with hyperdrive using amount.
+    /// @param _asBase Whether to use hyperdrive's base asset.
+    /// @param _amount Amount of assets to spend.
+    /// @return maturityTime Maturity timestamp of the opened position.
+    /// @return bondAmount Amount of bonds received.
+    function openLong(
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:38]: It would be good to add an extra data field to this for future proofing. We currently don't use this, but for yield sources that require some kind of KYC or off-chain attestation (some examples of this are RWA protocols like Ondo), will require extra data to be passed.

File: contracts/libraries/HyperdriveExecution.sol [hunk 10]
```diff
@@ -0,0 +1,578 @@
...
+        IHyperdrive self,
+        bool _asBase,
+        uint256 _amount
+    ) internal returns (uint256 maturityTime, uint256 bondAmount) {
+        (maturityTime, bondAmount) = self.openLong(
+            _amount,
+            0,
+            0,
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:46]: We might add a `TODO` here. I could see it being useful to have slippage/negative interest protection, but I could be missing why we don't want to do this here.

File: contracts/libraries/HyperdriveExecution.sol [hunk 11]
```diff
@@ -0,0 +1,578 @@
...
+            bondReservesDelta,
+            _vaultSharePrice,
+            spotPrice,
+            poolConfig.fees.curve,
+            poolConfig.fees.governanceLP
+        );
+        return bondReservesDelta;
+    }
+
+    /// @dev Calculate the fees involved with opening the long and apply them.
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:132]: This looks great! One thought is that if we're not using the other return values (`shareReservesDelta` and `totalGovernanceFee`), we can probably remove those. Wdyt?

File: contracts/libraries/HyperdriveExecution.sol [hunk 12]
```diff
@@ -0,0 +1,578 @@
...
+        }
+        return shareProceeds;
+    }
+
+    /// @dev Calculates the amount of output assets received from closing a
+    ///      long. The process is as follows:
+    ///        1. Calculates the raw amount using yield space.
+    ///        2. Subtracts fees.
+    ///        3. Accounts for negative interest.
+    ///        4. Converts to shares and back to account for any rounding issues.
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:239]: This is good. Might be good to add this to `_calculateOpenLong`.

File: contracts/libraries/HyperdriveExecution.sol [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+            _spotPrice,
+            _vaultSharePrice
+        );
+
+        // Calculate the number of shares to add to the shareReserves.
+        // shareReservesDelta, _shareAmount and totalGovernanceFee
+        // are all denominated in shares:
+        //
+        // shares = shares - shares
+        _shareReservesDelta -= totalGovernanceFee;
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:186]: I could be missing where these are used, but if we're not going to use these, we might as well not calculate them. I get that it might be helpful to keep this around for future PRs, but adding a TODO as a reminder would be a good idea.

File: contracts/libraries/HyperdriveExecution.sol [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+    // │ Close Long                                              │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @notice Close a long with the input position's bond amount and maturity.
+    /// @param _asBase Whether to receive hyperdrive's base token as output.
+    /// @param _position Position information used to specify the long to close.
+    /// @return proceeds The amount of output assets received from closing the long.
+    function closeLong(
+        IHyperdrive self,
+        bool _asBase,
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:201]: Same here with `extraData`. Would be good for future proofing.

File: contracts/libraries/HyperdriveExecution.sol [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+
+        // NOTE: Round down to underestimate the governance curve fee.
+        //
+        // We leave the governance fee in terms of bonds:
+        // governanceCurveFee = curve_fee * phi_gov
+        //                    = bonds * phi_gov
+        governanceCurveFee = curveFee.mulDown(_governanceLPFee);
+    }
+
+    /// @notice Obtains the vaultSharePrice from the hyperdrive instance.
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:470]: Since this function (and the others below are internal), they aren't part of a public interface, and it would make sense to change the `@notice`s to `@dev`s. In the [NatSpec guide](https://docs.soliditylang.org/en/latest/natspec-format.html), `@notice` is defined as containing "Explain to an end user what this does," and in the case of internal functions, the end user is the developer.

File: contracts/libraries/HyperdriveExecution.sol [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+uint256 constant HYPERDRIVE_LONG_EXPOSURE_LONGS_OUTSTANDING_SLOT = 3;
+uint256 constant HYPERDRIVE_SHARE_ADJUSTMENT_SHORTS_OUTSTANDING_SLOT = 4;
+
+/// @author DELV
+/// @title HyperdriveExecutionLibrary
+/// @notice Library to handle the execution of trades with hyperdrive.
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+library HyperdriveExecutionLibrary {
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:24]: Nit: Just as a general comment, some of these functions are prefixed with a leading underscore, and some aren't. We typically avoid the leading underscore in libraries. Any reason to include them here?

File: contracts/Everlong.sol (hunk 5 above)
Comment (mcclurejt) [reply contracts/Everlong.sol:332]: Yep I commented a list of todos on the "Everlong" issue, not final by any means but should overview what is to come

File: contracts/libraries/HyperdriveExecution.sol (hunk 16 above)
Comment (mcclurejt) [reply contracts/libraries/HyperdriveExecution.sol:24]: The meaning is consistent with what an underscore indicates in contracts, that the function is not to be called by external parties and is only used internally.

If this is against the desired style, is there another preferred method for indicating the same?

File: contracts/libraries/HyperdriveExecution.sol (hunk 10 above)
Comment (mcclurejt) [reply contracts/libraries/HyperdriveExecution.sol:46]: With regards to slippage, I was under the impression from our conversation earlier in the week that there is no slippage in an atomic scenario, and that the parameter is intended for MEV protection. Can you help clarify this?

Will add a todo wrt vaultshareprice.

File: contracts/libraries/HyperdriveExecution.sol (hunk 16 above)
Comment (jalextowle) [reply contracts/libraries/HyperdriveExecution.sol:24]: You should be able to make it a private function.

Comment (jalextowle) [reply contracts/libraries/HyperdriveExecution.sol:24]: That said, it’s not a big deal. I’m asking because it’s not what we usually do, but your reasoning makes sense

File: contracts/libraries/HyperdriveExecution.sol (hunk 15 above)
Comment (mcclurejt) [reply contracts/libraries/HyperdriveExecution.sol:470]: The intent with notice/dev usage in the library was to mimic their meanings within a contract context. As you and natspec mention, these are for external users, who for a library are developers.

That being said, natspec also says dev comments are to "Explain to a developer any extra details" yet our usage of them mostly involves high level descriptions of internal functions, with the details handled by block commenting within the function.

Regardless, I will do as you have asked and migrate all top level comments to @dev tags

File: contracts/libraries/HyperdriveExecution.sol (hunk 10 above)
Comment (jalextowle) [reply contracts/libraries/HyperdriveExecution.sol:46]: Good question, here's how I'd frame this. There is no slippage from the time of valuing positions to the time of opening the trade. That said, from the perspective of someone that wants to withdraw from Everlong, there may be slippage between sending out the transaction and the transaction actually being executed. I'm envisioning that this would be a protection for the user that is withdrawing. We could also implement this higher up in the stack though with some kind of `minimumPricePerShare`, which may make more sense in this context.

File: contracts/libraries/HyperdriveExecution.sol (hunk 15 above)
Comment (jalextowle) [reply contracts/libraries/HyperdriveExecution.sol:470]: I hear you. The way that I've treated this in the past is that `@notice` is for things that are part of the external interface, but I get what you are saying.

File: contracts/interfaces/IEverlong.sol [hunk 17]
```diff
@@ -1,31 +1,45 @@
...
-    IEverlongPositions
+    IEverlongPortfolio
 {
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Structs                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /**
+     * @notice Contains the information needed to identify an open Hyperdrive position.
+     */
```
Comment (jalextowle) [inline contracts/interfaces/IEverlong.sol:21]: Nit: Let's stick with the `///` style for multiline comments.

```suggestion
     /// @notice Contains the information needed to identify an open Hyperdrive position.
```

File: contracts/Everlong.sol [hunk 18]
```diff
@@ -66,64 +67,314 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ ERC4626                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @notice Calculate the total amount of assets controlled by everlong.
+    /// @notice To do this efficiently, the weighted average maturity is used.
+    /// @dev Underestimates the actual value by overestimating the average
+    ///      maturity of the portfolio.
+    /// @return Total amount of assets controlled by Everlong.
+    function totalAssets() public view override returns (uint256) {
```
Comment (jalextowle) [inline contracts/Everlong.sol:192]: Nice, this is super clean.

File: contracts/Everlong.sol [hunk 19]
```diff
@@ -66,64 +67,314 @@ import { EVERLONG_KIND, EVERLONG_VERSION } from "./libraries/Constants.sol";
...
+    /// @return output Total output received from closed positions.
+    function _closePositions(
+        uint256 _targetOutput
+    ) internal returns (uint256 output) {
+        while (!_portfolio.isEmpty() && output < _targetOutput) {
+            output += _hyperdrive.closeLong(_asBase, _portfolio.head());
+            _portfolio.handleClosePosition();
+        }
+        return output;
     }
```
Comment (jalextowle) [inline contracts/Everlong.sol:279]: These are super clean too. Nice!

File: contracts/libraries/HyperdriveExecution.sol [hunk 20]
```diff
@@ -0,0 +1,578 @@
...
+        //
+        // We calculate the flat portion of the governance fee as:
+        //
+        // governance_flat_fee = flat_fee * phi_gov
+        //                     = shares * phi_gov
+        //
+        // The totalGovernanceFee is the sum of the curve and flat governance fees.
+        totalGovernanceFee =
+            governanceCurveFee +
+            flatFee.mulDown(_governanceLPFee);
```
Comment (jalextowle) [inline contracts/libraries/HyperdriveExecution.sol:416]: Similar to a comment above, we probably don't need to calculate this if we're not going to use it.

**Review by jalextowle — COMMENTED** [review]
This is looking good. I left some comments, but it looks good to me once those are addressed or acknowledged.

**Review by jalextowle — APPROVED** [review]

File: contracts/Everlong.sol (hunk 3 above)
Comment (mcclurejt) [reply contracts/Everlong.sol:84]: Ahh when I had tested this I just looked at the testing function gas costs w/o checking the gas left. 

I guess still worth keeping for the lower deployment costs?

File: contracts/libraries/HyperdriveExecution.sol [hunk 21]
```diff
@@ -0,0 +1,578 @@
...
+            self.getPoolConfig().positionDuration
+        );
+
+        // Since we overestimate the time remaining to underestimate the
+        // proceeds, there is an edge case where _normalizedTimeRemaining > 1
+        // if the position was opened in the same checkpoint. For this case,
+        // we can just set it to 1e18.
+        if (timeRemaining > 1e18) {
+            timeRemaining = 1e18;
+        }
```
Comment (mcclurejt) [inline contracts/libraries/HyperdriveExecution.sol:520]: @jalextowle just highlighting this since it's kinda jank but gotta do it

File: contracts/Everlong.sol (hunk 3 above)
Comment (jalextowle) [reply contracts/Everlong.sol:84]: I don’t think so. It’s good practice for all immutables to have getters for better UX and DevEx (this helps with etching), and it seemed like most (all?) of these already had getters elsewhere. With that said, the cost will be the same, it’s just more verbose

Comment (mcclurejt) [reply contracts/Everlong.sol:84]: okay IIRC will need to leave some as internal to play nicely with solady erc4626 (don't see a need for us to reimplement and should speed up audit times not doing so), but bulk majority should be fine to make public

Comment (jalextowle) [reply contracts/Everlong.sol:84]: Yep, makes sense

File: contracts/libraries/HyperdriveExecution.sol (hunk 10 above)
Comment (mcclurejt) [reply contracts/libraries/HyperdriveExecution.sol:52]: added a todo

**Comment by mcclurejt** [comment]
> This is looking good. I left some comments, but it looks good to me once those are addressed or acknowledged.

going to fix test stuff, get review on that, then merge into this branch just fyi
